### PR TITLE
Added zookeeper multi-read support

### PR DESCRIPTION
**Pull request by asokol123**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

- [x] Implemented multi-read support for get operation
- [x] Implemented multi-read support for simple list operation
- [ ] Implemented server version check and fallback to loop-retry solution for old servers
- [x] Added multi-read support in ClickHouse Keeper

### CODE REVIEW

**Comment by antonio2368** [comment]
The `KeeperStorage` was refactored so it would be good to modify the Processors to be aligned with it
https://github.com/ClickHouse/ClickHouse/pull/37036

File: src/Common/ZooKeeper/ZooKeeperConstants.h [hunk 1]
```diff
@@ -51,6 +58,6 @@ static constexpr int32_t MAX_STRING_OR_ARRAY_SIZE = 1 << 28;  /// 256 MiB
 static constexpr int32_t DEFAULT_SESSION_TIMEOUT_MS = 30000;
 static constexpr int32_t DEFAULT_MIN_SESSION_TIMEOUT_MS = 10000;
 static constexpr int32_t DEFAULT_MAX_SESSION_TIMEOUT_MS = 100000;
-static constexpr int32_t DEFAULT_OPERATION_TIMEOUT_MS = 10000;
+static constexpr int32_t DEFAULT_OPERATION_TIMEOUT_MS = 100000;
```
Comment (alexey-milovidov) [inline src/Common/ZooKeeper/ZooKeeperConstants.h:61]: Why need this change?

File: src/Common/ZooKeeper/TestKeeper.cpp [hunk 2]
```diff
@@ -169,8 +194,22 @@ struct TestKeeperMultiRequest final : MultiRequest, TestKeeperRequest
...
+        else
+        {
+            op_kind = is_write;
+        }
+    }
+
     ResponsePtr createResponse() const override;
     std::pair<ResponsePtr, Undo> process(TestKeeper::Container & container, int64_t zxid) const override;
+
+    std::optional<bool> op_kind;
```
Comment (antonio2368) [inline src/Common/ZooKeeper/TestKeeper.cpp:212]: Use something like
```
enum class OpKind : uint8_t
{
    Read,
    Write
};
```
Code will be easier to read

File: src/Common/ZooKeeper/TestKeeper.cpp [hunk 3]
```diff
@@ -359,6 +398,22 @@ std::pair<ResponsePtr, Undo> TestKeeperSetRequest::process(TestKeeper::Container
...
+    req.path = path;
+    auto [full_response_ptr, undo] = req.process(container, zxid);
+    const auto * full_response = dynamic_cast<const ListResponse*>(full_response_ptr.get());
+    if (full_response == nullptr)
+    {
+        throw Exception("Logical error: expected to get ListResponse from ListRequest", Error::ZSYSTEMERROR);
+    }
+    SimpleListResponse response;
+    response.error = full_response->error;
+    response.names = full_response->names;
```
Comment (antonio2368) [inline src/Common/ZooKeeper/TestKeeper.cpp:413]:
```suggestion
    response.names = std::move(full_response->names);
```

File: src/Common/ZooKeeper/TestKeeper.cpp [hunk 4]
```diff
@@ -359,6 +398,22 @@ std::pair<ResponsePtr, Undo> TestKeeperSetRequest::process(TestKeeper::Container
...
+{
+    TestKeeperListRequest req;
+    req.path = path;
+    auto [full_response_ptr, undo] = req.process(container, zxid);
+    const auto * full_response = dynamic_cast<const ListResponse*>(full_response_ptr.get());
+    if (full_response == nullptr)
+    {
+        throw Exception("Logical error: expected to get ListResponse from ListRequest", Error::ZSYSTEMERROR);
+    }
+    SimpleListResponse response;
```
Comment (antonio2368) [inline src/Common/ZooKeeper/TestKeeper.cpp:411]: Why not instantly make a shared_ptr of `SimpleListResponse`?

File: src/Common/ZooKeeper/ZooKeeperConstants.h [hunk 5]
```diff
@@ -14,6 +14,12 @@ static constexpr XID PING_XID  = -2;
 static constexpr XID AUTH_XID  = -4;
 static constexpr XID CLOSE_XID = 0x7FFFFFFF;
 
+enum class OpKind : int32_t
+{
+    Read = 1,
+    Transaction = 2,
```
Comment (antonio2368) [inline src/Common/ZooKeeper/ZooKeeperConstants.h:20]: Why not write?

File: src/Common/ZooKeeper/ZooKeeperConstants.h [hunk 6]
```diff
@@ -14,6 +14,12 @@ static constexpr XID PING_XID  = -2;
 static constexpr XID AUTH_XID  = -4;
 static constexpr XID CLOSE_XID = 0x7FFFFFFF;
 
+enum class OpKind : int32_t
```
Comment (antonio2368) [inline src/Common/ZooKeeper/ZooKeeperConstants.h:17]:
```suggestion
enum class OpKind : uint8_t
```

File: utils/zookeeper-test/main.cpp [hunk 7]
```diff
@@ -70,6 +72,53 @@ void testCreateList(zkutil::ZooKeeper & zk)
     }
 }
 
+template<typename T, typename U>
+void assert_eq(const T& left, const U& right)
```
Comment (antonio2368) [inline utils/zookeeper-test/main.cpp:76]:
```suggestion
void assert_eq(const T & left, const U & right)
```

File: utils/zookeeper-test/main.cpp [hunk 8]
```diff
@@ -70,6 +72,53 @@ void testCreateList(zkutil::ZooKeeper & zk)
...
+{
+    if (left != right)
+    {
+        throw std::runtime_error("assert_eq failed");
+    }
+}
+
+template<typename T, typename U>
+T* assert_dynamic_cast(U* ptr)
```
Comment (antonio2368) [inline utils/zookeeper-test/main.cpp:85]:
```suggestion
T * assert_dynamic_cast(U * ptr)
```

**Comment by antonio2368** [comment]
Continued in #41410

**Comment by antonio2368** [comment]
Continued in #41410

**Pull request closed without merging**